File: 148.排序链表.go
/*
 * @lc app=leetcode.cn id=148 lang=golang
 *
 * [148] 排序链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func sortList(head *ListNode) *ListNode {
	return sortSubList(head, nil)
}

func sortSubList(start, end *ListNode) *ListNode {
	if start == nil {
		return start
	}
	if start.Next == end {
		start.Next = nil
		return start
	}
	slow, fast := start, start
	for fast != end {
		slow = slow.Next
		fast = fast.Next
		if fast != end {
			fast = fast.Next
		}
	}
	mid := slow
	return merge(sortSubList(start, mid), sortSubList(mid, end))
}

func merge(l1, l2 *ListNode) *ListNode {
	tmp := &ListNode{}
	tmp1 := tmp
	for l1 != nil && l2 != nil {
		if l1.Val <= l2.Val {
			tmp.Next = l1
			l1 = l1.Next
		} else {
			tmp.Next = l2
			l2 = l2.Next
		}
		tmp = tmp.Next
	}
	if l1 == nil {
		tmp.Next = l2
	} else {
		tmp.Next = l1
	}
	return tmp1.Next
}

// @lc code=end

File: C++/148.排序链表.cpp
/*
 * @lc app=leetcode.cn id=148 lang=cpp
 *
 * [148] 排序链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* sortList(ListNode* head) {
        return sortList(head, NULL);
    }

    ListNode* sortList(ListNode *start, ListNode *end) {
        if(start == NULL) return NULL;
        if(start->next == end) {
            start->next = NULL;
            return start;
        }
        ListNode* slow = start, *fast = start;
        while(fast != end) {
            slow = slow->next;
            fast = fast->next;
            if(fast != end) {
                fast = fast->next;
            }
        }
        ListNode* mid = slow;
        return mergeList(sortList(start, mid), sortList(mid, end));
    }
    ListNode* mergeList(ListNode* l1, ListNode* l2) {
        ListNode pre;
        ListNode *tmp = &pre, *tmp1 = l1, *tmp2 = l2;
        while(tmp1 != NULL && tmp2 != NULL) {
            if(tmp1->val < tmp2->val) {
                tmp->next = tmp1;
                tmp1 = tmp1->next;
            } else {
                tmp->next = tmp2;
                tmp2 = tmp2->next;
            }
            tmp = tmp->next;
        }
        if(tmp1 == NULL) {
            tmp->next = tmp2;
        } else {
            tmp->next = tmp1;
        }
        return pre.next;
    }
};
// @lc code=end

